[3기-C 권성준] springboot-url-shortener 과제 제출합니다.

Seongjun-Kwon opened the pull request:
## 📌 과제 설명
공태현(@rhdtn311 ) 님과 페어 프로그래밍을 통해 구현하였습니다.
코드리뷰 잘 부탁드리겠습니다! 😁

## 👩‍💻 요구 사항
- [x] URL 입력폼 제공 및 결과 출력
- [x] URL Shortening Key는 8 Character 이내로 생성
- [x] 단축된 URL 요청시 원래 URL로 리다이렉트
- [x] 단축된 URL에 대한 요청 수 정보저장 (optional)
- [ ] Shortening Key를 생성하는 알고리즘 2개 이상 제공하며 애플리케이션 실행중 동적으로 변경 가능 (optional)
---
## 👩‍💻 구현 내용
### API 문서
#### 1. URL 입력폼 화면 제공
| 메서드 | URL
:---: | :---: |
GET | / 
---
#### 2. 단축된 URL 생성
| 메서드 | URL
:---: | :---:
POST | /shortener

| Response Parameter | 타입 | 설명
:---: | :---: | :---:
originUrl | String | 원래 URL
shortUrl | String | 단축된 URL
requestCount | Number | 요청 수
---
#### 3. 생성된 상세 결과 페이지 제공 (원래 URL, 단축된 URL, 요청 수)
| 메서드 | URL
:---: | :---: 
GET | /shortener

| Request Parameter | 타입 | 설명
:---: | :---: | :---:
originUrl | String | 원래 URL
shortUrl | String | 단축된 URL
requestCount | Number | 요청 수
---
#### 4. 단축된 URL 로 요청 시 원래 URL 로 리다이렉트
| 메서드 | URL
:---: | :---:
GET | /{shortUrl}

| Path Parameter | 타입 | 설명
:---: | :---: | :---:
shortUrl | String | 단축된 URL
---
### DB 테이블
| 컬럼 명 | 타입 | NULL 가능 | 키 | 설명
:---: | :---: | :---: | :---: | :---:
id | bigint | x | pk | ID 값
algorithm | varchar(20) | x | uk| 알고리즘
origin_url | varchar(2000) | x | uk | 원래 URL
short_url | varchar(7) | o | - |단축된 URL
request_count | bigint | x | - | 요청 수
---
### 동작 화면

#### URL 입력폼 화면

![image](https://user-images.githubusercontent.com/68289543/210596993-25fbb2b4-a38b-4087-b5f1-415c6a27cde9.png)

#### URL 입력폼 화면 - 단축된 URL 생성 요청

![image](https://user-images.githubusercontent.com/68289543/210597113-3699b954-aacf-462f-a0c8-2142e7b50a5c.png)

#### 단축된 URL 생성 요청 후 상세 결과 페이지

![image](https://user-images.githubusercontent.com/68289543/210597413-b539d311-b3ca-4b4c-b58e-60850e78bf13.png)

#### 상세 결과 페이지 - 단축된 URL 클릭 시 원본 URL 로 이동

![gif-1](https://user-images.githubusercontent.com/68289543/210597475-b834e773-a70b-4e09-8223-6dab364fed13.gif)

#### 이미 DB에 존재하는 URL 인 경우 요청 수 증가

![git-2](https://user-images.githubusercontent.com/68289543/210597554-3629dc02-82dd-4671-9f61-7579abe84338.gif)

#### 잘못된 URL 을 입력하는 경우 사용자에게 오류 표시

![GIF-3](https://user-images.githubusercontent.com/68289543/210597608-88e5503d-7054-4205-b488-5294c24675c3.gif)

#### DB 에 없는 단축된 URL 요청을 한 경우 에러 페이지 이동

![gif-4](https://user-images.githubusercontent.com/68289543/210597617-666ca514-20e2-4d31-b0ec-798750d9aad8.gif)
---

Seongjun-Kwon committed:
chore: 프로젝트 생성

rhdtn311 committed:
chore : 프로젝트 초기 설정

rhdtn311 committed:
feat: ShortUrl 엔티티 추가

Seongjun-Kwon committed:
refactor: Entity 수정.

- Entity 명 수정: ShortUrl -> Url
- algorithm 필드 enum 타입으로 관리
- 요청 횟수 증가 메서드 추가

Seongjun-Kwon committed:
feat: UrlRepository 구현.

Seongjun-Kwon committed:
feat: 초기 화면 구현.

Seongjun-Kwon committed:
feat: Url 관련 DTO 구현.

rhdtn311 committed:
refactor : Url 엔티티 관련 수정

- Algorithm enum 삭제
- Url 엔티티 제약 조건, setShortUrl 추가

rhdtn311 committed:
chore : 패키지 구조 변경

rhdtn311 committed:
feat : Url short Base62 알고리즘 구현

rhdtn311 committed:
feat : short url 생성 기능 추가

rhdtn311 committed:
feat : short url 생성 기능 추가

Seongjun-Kwon committed:
feat: Short Url 상세 페이지 조회, Origin Url 로 리다이렉트하는 기능 구현.

rhdtn311 committed:
fix : 서비스 레이어에 프로토콜 제거된 URL을 보내지 않던 문제 해결

rhdtn311 committed:
feat : 구현한 기능에 맞는 화면단 구현

Seongjun-Kwon committed:
fix: short URL 눌러서 origin URL 로 리다이렉트 시 생기는 문제 해결.

- origin Url 에 / 같은 특수기호 있을 시 RedirectAttributes 에서 값을 제대로 담지 못함.

Seongjun-Kwon committed:
refactor: 사용자에게 보여주는 short URL 에서 프로토콜 정보 제거.

Seongjun-Kwon committed:
refactor: 홈 화면에 로고 추가.

Seongjun-Kwon committed:
feat: 404, 5xx 에러 페이지 추가.

Seongjun-Kwon committed:
feat: 잘못된 URL 검증하는 custom 어노테이션 구현.

rhdtn311 committed:
refactor : 잘못된 URL 입력 시 사용자 화면에 표시해줌

rhdtn311 committed:
refactor : ConstraintViolationException 예외 처리 추가

Seongjun-Kwon committed:
test: @UrlValid 테스트 추가.

Seongjun-Kwon committed:
test: URL 검증 테스트에 입력값으로 blank, empty 값 들어온 경우 테스트 추가.

rhdtn311 committed:
fix : Base62Algorithm decode() 메소드 수정

rhdtn311 committed:
test : Base62Algorithm encode(), decode() 테스트 코드 추가

Seongjun-Kwon committed:
test: UrlCreateDto 클래스 테스트 추가.

Seongjun-Kwon committed:
Merge branch 'main' of https://github.com/Marco-study/springboot-url-shortener

Seongjun-Kwon committed:
test: UrlServiceRequestDto 클래스 테스트 추가.

rhdtn311 committed:
refactor : shortURL 상세페이지로 리다이렉트 시 반환값에 Url Parameter 값 입력 제거

Seongjun-Kwon committed:
test: Url 엔티티 테스트 추가.

Seongjun-Kwon committed:
Merge branch 'main' of https://github.com/Marco-study/springboot-url-shortener

rhdtn311 committed:
test : UrlController 테스트 코드 작성

rhdtn311 committed:
test : 통합 테스트 코드 작성

- 새로운 Short URL 생성 성공
- 이미 존재하는 Orgin URL 인 경우 DB 에서 꺼내와 RequestCount + 1 성공
- shortURL 을 받아서 UrlResponseDto 로 반환 성공
- shortURL 을 받았는데 DB에 관련된 Url 이 존재하지 않는 경우 EntityNotFoundException 예외 발생

Seongjun-Kwon committed:
test: 필요 없는 테스트 삭제.

Seongjun-Kwon committed:
test: 통합 테스트 패키지 변경.

Seongjun-Kwon committed:
test: UrlService 테스트 구현.

Seongjun-Kwon committed:
test: 통합 테스트 수정.

Seongjun-Kwon committed:
refactor: 매직 넘버 제거.

Seongjun-Kwon committed:
docs: README 수정.

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/algorithm/Base62Algorithm.java:
```diff
@@ -0,0 +1,50 @@
+    private static final int SHORT_URL_LENGTH = 7;
+
+    public String encode(Long id) {
+        StringBuilder sb = new StringBuilder();
```
sb보단 더 의미는 단어를 사용하시면 어떨까요?

Seongjun-Kwon replied:
인코딩된 문자를 붙여주는 역할을 하기에 encodedTokenAppender 로 네이밍했습니다.

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/algorithm/Base62Algorithm.java:
```diff
@@ -0,0 +1,50 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class Base62Algorithm {
```
Algorithm이 변경될수도 있으니, 인터페이스로 확장성을 지켜보면 어떨까요?

Seongjun-Kwon replied:
수정했습니다!

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/algorithm/Base62Algorithm.java:
```diff
@@ -0,0 +1,50 @@
+            int paddingCount = SHORT_URL_LENGTH - shortUrl.length();
+
+            for (int i = 0; i < paddingCount; i++) {
+                sb.append("A");
```
'A'는 어떤 의미일까요?
상수화 시켜주면 좋을거 같아요.

Seongjun-Kwon replied:
패딩에 사용되는 문자이기에 PADDING_TOKEN 이름으로 상수화했습니다.

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/Url.java:
```diff
@@ -0,0 +1,60 @@
+@Table(
+        name = "url",
+        uniqueConstraints = {
+                @UniqueConstraint(
+                        name = "uk",
+                        columnNames = {"algorithm", "origin_url"}
+                )
```
👍

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/Url.java:
```diff
@@ -0,0 +1,60 @@
+    @Column(name = "short_url", length = 7)
+    private String shortUrl;
+
+    @Column(name = "request_count")
+    private Long requestCount;
```
이거 하다가 locking 걸릴수도 있을거 같은데요?

Seongjun-Kwon replied:
수정했습니다!
service 레이어 하나를 더 만들어서 해결해봤는데 괜찮은 방식인지 궁금합니다!

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/dto/UrlServiceRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+		return Url.builder()
+			.originUrl(originUrl)
+			.algorithm(algorithm)
+			.requestCount(1L)
```
dto가 아니라, entity에 default 값으로 지정해두면 어때요?

Seongjun-Kwon replied:
수정했습니다. 
디폴트 값을 정해주는 것을 @Column 의 columnDefinition 속성으로 넣어주어서 DB 레벨에서 넣을지, 아니면 코드 레벨에서 생성자에 넣어줄지가 고민이 되었습니다.
jpa 의 auto ddl 옵션을 쓰지 않는 경우도 있으니 코드 레벨에서 생성자에 디폴트 값을 넣어주는 방식을 이용했습니다.

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/dto/UrlCreateDto.java:
```diff
@@ -0,0 +1,26 @@
+	}
+
+	private String removeProtocolFromOriginUrl() {
+		if (this.originUrl.startsWith("https://")) {
+			return this.originUrl.replace("https://", "");
+		}
+		if (this.originUrl.startsWith("http://")) {
+			return this.originUrl.replace("http://", "");
+		}
```
이겨있는 값들도 상수화 👍

Seongjun-Kwon replied:
"http://" -> HTTP_PROTOCOL
"https://" -> HTTPS_PROTOCOL
"" -> 이거는 "" 자체가 더 직관적으로 보여서 수정하지 않았습니다!

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/dto/UrlCreateDto.java:
```diff
@@ -0,0 +1,26 @@
+
+	private String removeProtocolFromOriginUrl() {
+		if (this.originUrl.startsWith("https://")) {
+			return this.originUrl.replace("https://", "");
```
DTO에서 이런걸 하는게 맞을까요?

Seongjun-Kwon replied:
dto 는 데이터 전송만 담당하고 값을 수정하는 책임을 가지는 건 옳지 않은 것 같습니다.
사용자에게 받은 URL 에서 프로토콜 정보를 제거하는 로직은 엔티티가 가지도록 변경했습니다.

추가적으로 인코딩된 AAAAAAB 같은 값에 도메인 정보를 붙여서 사용자가 사용할 수 있는 localhost:8080/AAAAAAB 로 바꾸는 로직은 도메인 정보가 변경될 수 있으므로 프론트 측이 하는 것이 좋다고 생각하여 옮겼습니다!

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/UrlService.java:
```diff
@@ -0,0 +1,54 @@
+	@Transactional
+	public UrlResponseDto createShortUrl(UrlServiceRequestDto urlRequestDto) {
+
+		Optional<Url> optionalUrl = urlRepository.findByOriginUrl(urlRequestDto.originUrl());
```
OriginalUrl을 어느곳에서도 저장하지 않는데, 어떻게 DB에서 조회하죠?

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/UrlService.java:
```diff
@@ -0,0 +1,54 @@
+
+		Optional<Url> optionalUrl = urlRepository.findByOriginUrl(urlRequestDto.originUrl());
+
+		if (optionalUrl.isPresent()) {
+			Url url = optionalUrl.get();
+			optionalUrl.get().increaseRequestCount();
+
+			return new UrlResponseDto(url.getOriginUrl(), url.getShortUrl(), url.getRequestCount());
```
오오 중복조회 👍

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/UrlService.java:
```diff
@@ -0,0 +1,54 @@
+		}
+
+		Url url = urlRequestDto.toEntity();
+		Url savedUrl = urlRepository.save(url);
+		String shortUrl = base62Algorithm.encode(savedUrl.getId());
+		savedUrl.setShortUrl(shortUrl);
```
한번에 Insert는 못하나요?

Seongjun-Kwon replied:
현재 로직에서 auto increment 로 생성된 id 값을 인코딩해서 short URL 을 만듭니다.
그렇기에 shortURL 을 업데이트 해주는 로직이 필요합니다.
만약 DB 에서 생성된 값을 이용하지 않는다면 한번에 insert 가 가능할 것 같습니다.

ksy90101 commented on src/main/java/com/programmers/springbooturlshortener/domain/url/util/UrlValid.java:
```diff
@@ -0,0 +1,21 @@
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = UrlValidation.class)
+public @interface UrlValid {
```
멋지네요 👍

ksy90101 reviewed (changes requested):
잘 구현했네요 👍 
피드백만 확인해주세요~

Seongjun-Kwon committed:
chore: 패키지 구조 변경.

Seongjun-Kwon committed:
refactor: UrlServiceResponseDto 를 만들어서 컨트롤러에서 사용하는 DTO 에 서비스가 의존하지 않도록 함.

Seongjun-Kwon committed:
refactor: 단축된 URL 생성 후 상세 페이지로 리다이렉트 시 데이터들을 FlashMap 으로 전송하게 함.

Seongjun-Kwon committed:
refactor: 모호한 변수명 수정.

Seongjun-Kwon committed:
refactor: 알고리즘 확장성을 고려하여 Algorithm 인터페이스 생성.

Seongjun-Kwon committed:
refactor: Url 생성 시 requestCount 디폴트 값을 엔티티에서 넣어주게 함.

Seongjun-Kwon committed:
refactor: Service 에서 UrlServiceResponseDto 의 requestCount 필드에 1L 을 직접 넣던 코드 수정.

Seongjun-Kwon committed:
refactor: 필요 없는 어노테이션 제거.

Seongjun-Kwon committed:
refactor: URL 파싱 로직의 위치 변경.

- 기존에 DTO 가 가지던 사용자에게 입력 받은 URL 에서 프로토콜 정보를 제거하는 로직을 엔티티가 가지게 함.

- 인코딩된 값에 도메인 정보를 붙여서 사용자가 사용할 수 있는 단축된 URL 로 만드는 작업을 프론트에서 하도록 함.

Seongjun-Kwon committed:
chore: 패키지 의존성을 위해 패키지 구조 변경.

Seongjun-Kwon committed:
refactor : requestCount 증가 시 동시성 문제 해결

ksy90101 commented:
고생했어요.
